from passlib.context import CryptContext
import os
from fastapi import HTTPException
from app.models import models
from jose import JWTError, jwt

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

def hash_password(password: str) -> str:
    return pwd_context.hash(password)

def verify_password(plain_password: str, hashed_password: str) -> bool:
    print(f"Verificando senha: {plain_password} contra hash: {hashed_password}")  
    return pwd_context.verify(plain_password, hashed_password)


SECRET_KEY = os.getenv("SECRET_KEY", "mysecretkey")
ALGORITHM = "HS256"

def create_session_token(user: models.User) -> str:
    payload = {
        "user_id": user.id,
        "email": user.email,
        "role": user.role.name,
        "username": user.username 
    }
    return jwt.encode(payload, SECRET_KEY, algorithm=ALGORITHM)

def verify_session_token(token: str) -> dict:
    try:
        data = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
    except JWTError:
        raise HTTPException(status_code=401, detail="Token inv√°lido ou expirado")
    return data
